pkg/controller/servicebindingrequest/annotations/resource.go

```diff
@@ -0,0 +1,200 @@
+//
+// In the case the resource reference and source path are the same, the resource reference path is
+// returned; otherwise the output path is the source path prefixed by the resource reference path.
+func getOutputPath(bindingInfo *BindingInfo) string {
+	outputPath := bindingInfo.ResourceReferencePath
+	if bindingInfo.ResourceReferencePath != bindingInfo.SourcePath {
+		outputPath = bindingInfo.ResourceReferencePath + "." + bindingInfo.SourcePath
+	}
+	return outputPath
+}
+
+// getInputPathFields infers the input path fields based on the given bindingInfo value.
+//
+// In the case the resource reference path and source path are the same and no input path prefix has
+// been given, an empty slice is returned.
+//
+// In the case inputPathPrefix is present, it is prepended to the resulting slice.
+//
+// In the case the resource reference and source paths are different, the source path is appended to
+// the resulting slice.
+func getInputPathFields(bindingInfo *BindingInfo, inputPathPrefix *string) []string {
+	inputPathFields := []string{}
+	if bindingInfo.ResourceReferencePath != bindingInfo.SourcePath {
+		inputPathFields = append(inputPathFields, bindingInfo.SourcePath)
+	}
+	if inputPathPrefix != nil && len(*inputPathPrefix) > 0 {
+		inputPathFields = append([]string{*inputPathPrefix}, inputPathFields...)
+	}
+	return inputPathFields
+}
+
+// Handle returns the value for an external resource strategy.
+func (h *ResourceHandler) Handle() (Result, error) {
+	name, err := h.discoverRelatedResourceName()
+	if err != nil {
+		return Result{}, err
+	}
+
+	ns := h.resource.GetNamespace()
+	resource, err := h.
+		client.
+		Resource(h.relatedGroupVersionResource).
+		Namespace(ns).
+		Get(name, metav1.GetOptions{})
+	if err != nil {
+		return Result{}, fmt.Errorf("error handling annotation: %w", err)
+	}
+
+	inputPathFields := getInputPathFields(h.bindingInfo, h.inputPathRoot)
+	val, ok, err := unstructured.NestedFieldCopy(resource.Object, inputPathFields...)
+	if !ok {
+		return Result{}, InvalidArgumentErr(strings.Join(inputPathFields, ", "))
+	}
+	if err != nil {
+		return Result{}, err
+	}
+
+	if mapVal, ok := val.(map[string]interface{}); ok {
+		tmpVal := make(map[string]interface{})
+		for k, v := range mapVal {
+			decodedVal, err := h.valueDecoder(v)
+			if err != nil {
+				return Result{}, err
+			}
+			tmpVal[k] = decodedVal
+		}
+		val = tmpVal
+	} else {
+		val, err = h.valueDecoder(val)
+		if err != nil {
+			return Result{}, err
+		}
+	}
+
+	typ, err := discoverBindingType(h.bindingInfo.Value)
+	if err != nil {
+		return Result{}, err
+	}
+
+	outputPath := getOutputPath(h.bindingInfo)
+
+	return Result{
+		Object: nested.ComposeValue(val, nested.NewPath(outputPath)),
+		Type:   typ,
+		Path:   outputPath,
+	}, nil
+}
+
+// stringValueDecoder asserts the given value 'v' and returns its string value.
+func stringValueDecoder(v interface{}) (string, error) {
```

Comment:
Not so sure; this function is meant to fit in the `valueDecoder` member in `Resource`. 

Perhaps it could be renamed to `decodeStringValue`, but doesn't change the fact the function is there to fit this particular use case.

Reply:
the name of the function in `Resource` is not important, although that could be improved as well. `decode` implies that the returned string value will be somehow encoded in the given interface. It does not have to be like that. On a more abstract level, we just want to get a string representation for the given interface. IMHO, `stringValue` (with or without `OrError` suffix) is just fine.

Reply:
Updated.